### PR TITLE
#166 Refactored predict cubes

### PR DESCRIPTION
**Pull request by vessemer**
The `predict_cubes` function has been refactored to reduce its complexity. 

## Description
The `predict_cubes` is now splitted on the support functions:
-[`prepare_data`](https://github.com/concept-to-clinic/concept-to-clinic/compare/master...vessemer:166_predict_cubes_refactoring?expand=1#diff-632542ba71c59c64b64a4bc113a8c4b6R148), by a given patient ID it returns three np.ndarray.
-[`annotate`](https://github.com/concept-to-clinic/concept-to-clinic/compare/master...vessemer:166_predict_cubes_refactoring?expand=1#diff-632542ba71c59c64b64a4bc113a8c4b6R229), by a given model and a volumetric data, it returns a DataFrame.
-[`stats_from_batch`](https://github.com/concept-to-clinic/concept-to-clinic/compare/master...vessemer:166_predict_cubes_refactoring?expand=1#diff-632542ba71c59c64b64a4bc113a8c4b6R289), for each nodule in a batch it returns a list of DataFrame

## Reference to official issue
This addresses #166

## CLA
- [x] I have signed the CLA; if other committers are in the commit history, they have signed the CLA as well

### CODE REVIEW

File: prediction/src/algorithms/identify/prediction.py [hunk 1]
```diff
@@ -178,89 +215,108 @@ def predict_cubes(model_path, patient_id, magnification=1, ext_name=""):  # noqa
...
+        predict_volume (np.ndarray): a volumetric placeholder for nodule probability storage.
+        batch_list_coords (list[list[int]]): list of corresponding coordinates for each sample of a batch, in zyx order.
+        annotation_index (int): index in the general sequence.
+
+    Returns:
+        list[pandas.DataFrame] containing anno_index, coord_x, coord_y, coord_z, diameter, nodule_chance, diameter_mm
+        of s in a batch.
+    """
+
+    patient_predictions_csv = list()
```
Comment (lamby) [inline prediction/src/algorithms/identify/prediction.py]: ``[]`` is usually bettter

File: prediction/src/algorithms/identify/prediction.py [hunk 2]
```diff
@@ -178,89 +215,108 @@ def predict_cubes(model_path, patient_id, magnification=1, ext_name=""):  # noqa
...
+        batch_list_coords.append((z, y, x))
+        img_prep = prepare_image_for_net3D(cube_img)
+        batch_list.append(img_prep)
+        if len(batch_list) % BATCH_SIZE == 0:
+            batch_data = np.vstack(batch_list)
+            p = model.predict(batch_data, batch_size=BATCH_SIZE)
+            ppc, annotation_index = stats_from_batch(p, patient_img.shape, predict_volume,
+                                                     batch_list_coords, annotation_index)
+            patient_predictions_csv.extend(ppc)
+            batch_list = list()
```
Comment (lamby) [inline prediction/src/algorithms/identify/prediction.py]: (ditto)

File: prediction/src/algorithms/identify/prediction.py [hunk 3]
```diff
@@ -178,89 +215,108 @@ def predict_cubes(model_path, patient_id, magnification=1, ext_name=""):  # noqa
...
+    Returns:
+        pandas.DataFrame containing anno_index, coord_x, coord_y, coord_z, diameter, nodule_chance, diameter_mm
+        of each found nodule.
+    """
+
+    done_count = 0
+    skipped_count = 0
+    annotation_index = 0
+
+    batch_list = list()
```
Comment (lamby) [inline prediction/src/algorithms/identify/prediction.py]: (ditto)

File: prediction/src/algorithms/identify/prediction.py [hunk 4]
```diff
@@ -160,15 +199,13 @@ def predict_cubes(model_path, patient_id, magnification=1, ext_name=""):  # noqa
...
-    flip_ext = ""
-
-    dst_dir += "predictions" + str(int(magnification * 10)) + holdout_ext + flip_ext + "_" + ext_name + "/"
+    dst_dir += "predictions" + str(int(magnification * 10)) + "_" + ext_name + "/"
     if not os.path.exists(dst_dir):
         os.makedirs(dst_dir)
 
     model = get_net(input_shape=(CUBE_SIZE, CUBE_SIZE, CUBE_SIZE, 1),
                     load_weight_path=model_path)
+
```
Comment (lamby) [inline prediction/src/algorithms/identify/prediction.py]: Please make these changes in separate commits next time - very useful for speed up review!

File: prediction/src/algorithms/identify/prediction.py [hunk 5]
```diff
@@ -160,15 +199,13 @@ def predict_cubes(model_path, patient_id, magnification=1, ext_name=""):  # noqa
...
     patient_ids = [patient_id]
     for file_name in os.listdir(EXTRACTED_IMAGE_DIR):
         if not os.path.isdir(EXTRACTED_IMAGE_DIR + file_name):
```
Comment (lamby) [inline prediction/src/algorithms/identify/prediction.py]: Shouldn't we be using ``os.path.join`` ? :)

File: prediction/src/algorithms/identify/prediction.py (hunk 4 above)
Comment (vessemer) [reply prediction/src/algorithms/identify/prediction.py]: Ok, I'll try to get rid of this habit. Thank you for the review!

File: prediction/src/algorithms/identify/prediction.py [hunk 6]
```diff
@@ -160,107 +199,124 @@ def predict_cubes(model_path, patient_id, magnification=1, ext_name=""):  # noqa
...
-    dst_dir += "predictions" + str(int(magnification * 10)) + holdout_ext + flip_ext + "_" + ext_name + "/"
+    dst_dir = os.path.join(dst_dir, "predictions" + str(int(magnification * 10)) + "_" + ext_name)
     if not os.path.exists(dst_dir):
         os.makedirs(dst_dir)
 
     model = get_net(input_shape=(CUBE_SIZE, CUBE_SIZE, CUBE_SIZE, 1),
                     load_weight_path=model_path)
+
     patient_ids = [patient_id]
     for file_name in os.listdir(EXTRACTED_IMAGE_DIR):
```
Comment (lamby) [inline prediction/src/algorithms/identify/prediction.py]: listdir returns items in nondeterminstic order; doesn't this make a difference here?

Comment (vessemer) [reply prediction/src/algorithms/identify/prediction.py]: In the original  Julian de Wit [implementation](https://github.com/juliandewit/kaggle_ndsb2017/blob/6de658e233201e573f1a6b2d150e0d51eaae9927/step3_predict_nodules.py#L198) `os.listdir` was used. The order was not necessary there scince each `file_name` represents a different patient. In the [adopted version](https://github.com/WGierke/concept-to-clinic/blob/2036289144cce659561539e2f29861648ba4842b/prediction/src/algorithms/identify/prediction.py#L251), `return df` returns a data frame only for the last patient, though. Since it's not the [original behaviour](https://github.com/juliandewit/kaggle_ndsb2017/blob/6de658e233201e573f1a6b2d150e0d51eaae9927/step3_predict_nodules.py#L315) followed by [this](https://github.com/juliandewit/kaggle_ndsb2017/blob/6de658e233201e573f1a6b2d150e0d51eaae9927/step3_predict_nodules.py#L100), I've corrected that to store all data frames in a dictionary which will be returned.

Comment (lamby) [reply prediction/src/algorithms/identify/prediction.py]: Can you add this as a comment to the code itself? This (awesome!) comment will get lost here, alas … :(

**Comment by lamby** [comment]
LGTM otherwise!

File: prediction/src/algorithms/identify/prediction.py (hunk 6 above)
Comment (vessemer) [reply prediction/src/algorithms/identify/prediction.py]: I've added [the comment](https://github.com/concept-to-clinic/concept-to-clinic/pull/172/files#diff-632542ba71c59c64b64a4bc113a8c4b6R215) but I'm not completely sure that I've done it correctly in the code base.

**Comment by lamby** [comment]
>  I'm not completely sure that I've done it correctly in the code base.

What do you mean?

**Comment by lamby** [comment]
Merged, thanks!
